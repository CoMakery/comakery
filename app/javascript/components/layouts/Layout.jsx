import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import Icon from './../styleguide/Icon'

class Layout extends React.Component {
  goBack() {
    typeof window === 'undefined' ? null : window.history.back()
  }

  render() {
    const {
      className,
      title,
      hasBackButton,
      subfooter,
      sidebar,
      children,
      ...other
    } = this.props

    const classnames = classNames(
      'layout',
      className
    )

    return (
      <React.Fragment>
        <div className={classnames} {...other}>
          { hasBackButton &&
            <div className="layout--back-button">
              <Icon name="iconCloseCopy.svg" onClick={this.goBack} />
            </div>
          }

          <div className="layout--content">
            <div className="layout--content--title">
              {title}
            </div>

            <hr className="layout--content--hr" />

            <div className="layout--content--wrapper">
              { sidebar &&
                <div className="layout--content--sidebar">
                  {sidebar}
                </div>
              }

              <div className={sidebar ? 'layout--content--content__sidebared' : 'layout--content--content'}>
                {children}
              </div>
            </div>
          </div>

          { subfooter &&
            <div className="layout--subfooter">
              <div className="layout--subfooter--buttons">
                {subfooter}
              </div>
            </div>
          }
        </div>
      </React.Fragment>
    )
  }
}

Layout.propTypes = {
  className    : PropTypes.string,
  title        : PropTypes.string,
  hasBackButton: PropTypes.bool,
  subfooter    : PropTypes.object,
  sidebar      : PropTypes.object
}
Layout.defaultProps = {
  className    : '',
  title        : 'title',
  hasBackButton: false,
  subfooter    : null,
  sidebar      : null
}
export default Layout
