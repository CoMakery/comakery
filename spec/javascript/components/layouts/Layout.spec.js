import React from 'react'
import { shallow } from 'enzyme'
import Layout from 'components/layouts/Layout'

describe('Layout', () => {
  it('renders correctly without props', () => {
    const wrapper = shallow(<Layout />)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.exists('.layout')).toBe(true)
    expect(wrapper.exists('.layout--back-button')).toBe(false)
    expect(wrapper.exists('.layout--content')).toBe(true)
    expect(wrapper.exists('.layout--content--title')).toBe(true)
    expect(wrapper.exists('.layout--content--sidebar')).toBe(false)
    expect(wrapper.exists('.layout--content--content')).toBe(true)
    expect(wrapper.exists('.layout--content--content__sidebared')).toBe(false)
    expect(wrapper.exists('.layout--subfooter')).toBe(false)
  })

  it('renders correctly with custom title', () => {
    const wrapper = shallow(<Layout title="test" />)

    expect(wrapper.find('.layout--content--title').text()).toBe('test')
  })

  it('renders correctly with custom child', () => {
    const wrapper = shallow(<Layout><p>test</p></Layout>)

    expect(wrapper.find('.layout--content--content p').text()).toBe('test')
  })

  it('renders correctly with custom sidebar', () => {
    const wrapper = shallow(<Layout sidebar={<p>test</p>} />)

    expect(wrapper.find('.layout--content--sidebar p').text()).toBe('test')
    expect(wrapper.exists('.layout--content--content')).toBe(false)
    expect(wrapper.exists('.layout--content--content__sidebared')).toBe(true)
  })

  it('renders correctly with custom className', () => {
    const wrapper = shallow(<Layout className="__test" />)

    expect(wrapper.exists('.layout.__test')).toBe(true)
  })

  it('renders correctly with back button', () => {
    const wrapper = shallow(<Layout hasBackButton />)

    expect(wrapper.exists('.layout--back-button')).toBe(true)
  })
})
