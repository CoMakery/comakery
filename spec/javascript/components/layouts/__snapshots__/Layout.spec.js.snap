// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout renders correctly without props 1`] = `
<Fragment>
  <div
    className="layout"
  >
    <div
      className="layout--content"
    >
      <div
        className="layout--content--title"
      >
        title
      </div>
      <hr
        className="layout--content--hr"
      />
      <div
        className="layout--content--wrapper"
      >
        <div
          className="layout--content--content"
        />
      </div>
    </div>
  </div>
</Fragment>
`;
